lib/spack/spack/cmd/checksum.py

```diff
@@ -119,49 +130,113 @@ def checksum(parser, args):
+        print()
+        print(version_lines)
+        print()
+
+
+def verify_checksums(pkg, version_hashes, url_dict):
+    """
+    Verify checksums present in version_hashes against those present
+    in the package's instructions.
+
+    Args:
+        pkg (package): A package class for a given package in Spack.
+        version_hashes (dict): A dictionary of the form: version -> checksum.
+        url_dict (dict): A dictionary of the form: version -> URL.
+
+    """
+    results = []
+    num_verified = 0
+    failed = False
+
+    max_len = max(len(str(v)) for v in version_hashes)
+    num_total = len(version_hashes)
+
+    for version, sha in version_hashes.items():
+        if not version in pkg.versions:
+            msg = "No previous checksum"
+            status = "-"
+
+        elif sha == pkg.versions[version]["sha256"]:
+            msg = "Correct"
+            status = "="
+            num_verified += 1
 
-    if args.add_to_package:
-        filename = spack.repo.path.filename_for_package_name(pkg.name)
-        # Make sure we also have a newline after the last version
-        versions = [v + "\n" for v in version_lines.splitlines()]
-        versions.append("\n")
-        # We need to insert the versions in reversed order
-        versions.reverse()
-        versions.append("    # FIXME: Added by `spack checksum`\n")
-        version_line = None
-
-        with open(filename, "r") as f:
-            lines = f.readlines()
-            for i in range(len(lines)):
-                # Black is drunk, so this is what it looks like for now
-                # See https://github.com/psf/black/issues/2156 for more information
-                if lines[i].startswith("    # FIXME: Added by `spack checksum`") or lines[
-                    i
-                ].startswith("    version("):
-                    version_line = i
-                    break
-
-        if version_line is not None:
-            for v in versions:
-                lines.insert(version_line, v)
-
-            with open(filename, "w") as f:
-                f.writelines(lines)
-
-            msg = "opening editor to verify"
-
-            if not sys.stdout.isatty():
-                msg = "please verify"
-
-            tty.info(
-                "Added {0} new versions to {1}, "
-                "{2}.".format(len(versions) - 2, args.package, msg)
-            )
-
-            if sys.stdout.isatty():
-                editor(filename)
         else:
-            tty.warn("Could not add new versions to {0}.".format(args.package))
+            msg = sha
+            status = "x"
+            failed = True
+
+        results.append("{0:{1}}  {2} {3}".format(str(version), max_len, f"[{status}]", msg))
+
+    # Display table of checksum results.
+    tty.msg(f"Found {num_verified} of {num_total}", "", *llnl.util.lang.elide_list(results))
+
+    # Terminate at the end of function to prevent additional output.
+    if failed:
+        tty.die("Invalid checksums found.")
```

Comment:
If we `print()` before the `die()` I believe we won't get the pretty red error prefix. We could `error()` before `die()` but that feels like calling two functions instead of one that does the same thing.

Reply:
It does and it doesn't.  My suggestion was to add a blank line before the `==> Error:` for readability.  The extra call isn't a performance issue but it isn't a requirement either.

Reply:
Fixed! I agree this makes the error more readable.